### DISCUSSION TITLE
Built-in data warehouse

### DISCUSSION THREAD

**Discussion by codyebberson** [medplum/medplum#3011] [Ideas]
Many self-hosting Medplum customers have an existing data warehouse solution such as Snowflake or Databricks.  We work with enterprise customers to setup HTTP webhooks or Postgres data dumps to ETL the data.

On the other hand, many Medplum customers do not have this, for a variety of reasons.

I'm starting this discussion to explore options for what we could do in our CDK infra-as-code to take advantage of AWS services such as AWS Redshift or AWS Athena.

### Short term

There are 2 resource types in Medplum which can accumulate quickly and age-out quickly: `Login` and `AuditEvent`.

`Login` is a custom Medplum resource that represents a user authentication event, and effectively acts as a "user session".  When patients login via Medplum Auth, a self-hosted server can quickly accumulate many `Login` resources.  Once a `Login` has expired, it does not provide any operational value.  You may still want them for analytics or customer support purposes.

`AuditEvent` is a standard FHIR resource which has many purposes.  Developers can create their own `AuditEvent` resources for any purpose.  Medplum server automatically creates `AuditEvent` resources on bot invocations and webhook calls.  If your server heavily uses bots or subscriptions, you will accumulate many `AuditEvent` resources.  The operational value of an `AuditEvent` rapidly declines after 30-90 days.  However, your organization may require longer data retention for legal or compliance purposes.

For both `Login` and `AuditEvent`, Medplum provides built-in tools to purge resources older than a specified time period.  That feature should really only be used if those resources are stored in a data warehouse.

### Long term

If Medplum server had excellent support for seamlessly using Postgres and a data warehouse in harmony, it would open up many interesting possibilities.

For example, some organizations heavily rely on `Task` resources for workflow management.  In general, you always want those `Task` resources to be available (imagine looking at a patient profile when they come back for an appointment 2 years after their last).

### P1

1. Data warehouse is opt-in in Medplum CDK code
2. If enabled, on every Postgres write operation, also write the data warehouse

### P2

1. Add server option to auto purge `Login` and `AuditEvent` resources by configurable criteria

### P3

1. Supplemental implementation of Medplum server `Repository` class which allows limited FHIR read/search functionality from data warehouse as a data source
2. ??? Maybe an option to exclusively use data warehouse for `AuditEvent` ???

### P4

1. Seamless support for reading from both Postgres and the data warehouse

## Redshift vs Athena

The two obvious solutions for a built-in data warehouse would be [AWS Redshift]() and [AWS Athena]().

There are many articles that compare these services:
* https://blog.panoply.io/an-amazonian-battle-comparing-athena-and-redshift
* https://chartio.com/resources/tutorials/redshift-vs-athena/
* https://www.chaossearch.io/blog/when-to-deploy-aws-redshift-or-athena-use-cases

Main differences:
* Redshift pricing is by the hour, Athena pricing is by "data scanned during the query"
* Redshift will probably be more expensive, unless you use Athena a ton
* Redshift generally faster, Athena generally slower
* Redshift is older, and therefore more popular and more widely used, more 3rd party integrations

**Comment by codyebberson** [comment]
A few recent thoughts:

1. Consider SQL-on-FHIR
    * IG: https://build.fhir.org/ig/FHIR/sql-on-fhir-v2/
    * Standards are good, and it would be great to align with that initiative
    * However, it is very early, quite rough, and going through a lot of thrash
    * My recommendation would be to treat it as an informational data point, and not a requirement
2. Consider BigQuery
    * Built-in FHIR analytics: https://cloud.google.com/architecture/analyzing-fhir-data-in-bigquery
    * Some large Medplum customers have reported high satisfaction with BigQuery for analytics
    * We could embrace BigQuery, and build a Medplum-to-BigQuery adaptor or exporter
    * But that requires BigQuery and GCP, which most Medplum customers do not use
    * So instead, studying it and learning best practices might be more fruitful
3. Leaning towards Redshift over Athena
    * As we've used Athena more for internal analytics on HTTP requests and audit logs, we now better understand its pros/cons
    * The main limitation is that Athena "indexing" is limited to S3 object path
    * That means you searching on anything other than "lastUpdated" would be very slow and expensive
    * That's a pretty significant limitation, and pushes the needle toward Redshift

**Comment by codyebberson** [comment]
**Updated Thoughts (May 2025)**

1. **Expanded Scope Beyond Login & AuditEvent**
   Initially, the focus was on moving `Login` and `AuditEvent` resources to a warehouse for long-term storage and purging them from the operational Postgres database. However, our customers are increasingly requesting support for *all* FHIR resources. This signals the broader need for a robust analytics and data science platform that doesn’t limit itself to specific resource types.

2. **Lessons from Google FHIR Server**
   Google’s FHIR server and BigQuery integrations have emerged as a strong reference point for FHIR analytics. We don’t necessarily want to push Google FHIR as the sole solution, but there’s a lot to learn about:

   * How they flatten (or otherwise structure) FHIR data into a tabular form.
   * Query patterns they support out-of-the-box.
   * How they handle large-scale analytical workloads on complex FHIR data.

3. **Increasing Demand for Cloud-Agnostic Solutions**
   While Redshift and Athena remain compelling for customers on AWS, not all Medplum users are exclusively on AWS. Many are pushing for cloud-agnostic and open-source solutions. This suggests we should keep in mind:

   * **ClickHouse**: Known for high-performance analytics; an interesting option for FHIR data.
   * **Apache Spark**: Could be leveraged for distributed processing and flexible transformations.
   * **ELK Stack (Elasticsearch, Logstash, Kibana)**: Popular for log/audit data; can also support general analytics for certain use cases.
   * **Other data warehouses**: Snowflake, Databricks, etc.

4. **Future Considerations**

   * **SQL-on-FHIR Progress**: The HL7 SQL-on-FHIR spec is evolving, and while it’s still early, aligning with the standard could provide a consistent SQL-based approach across different platforms.
   * **Unified Repository**: Eventually, if the Medplum repository can route queries intelligently to either the operational Postgres cluster or an analytical store, we can achieve the best of both worlds.

Overall, our approach should be flexible and modular, allowing customers to integrate their data into any modern analytics platform—whether that’s AWS-native, Google BigQuery, or on-prem open-source tooling like ClickHouse or Spark.

**Comment by WELLaaronaston** [comment]
Another option is (realtime streaming to) Iceberg tables (files), which you could then mount in a data lake of your choice.  On AWS both Redshift and Athena seem to support Iceberg (according to the AWS documentation).  Might be a happy middle ground to allow customers to decide which query engine they want to use while keeping the plumbing agnostic.

**Reply by codyebberson** [reply]
Excellent call, @WELLaaronaston.  This would allow for a data lakehouse experience that offers features comparable to Google's FHIR+BigQuery analytics, such as fast SQL querying and near-real-time data availability, while using open standards.

Deep Research findings attached: [Design_ Cloud-Agnostic Data Warehouse Support in Medplum via Apache Iceberg.pdf](https://github.com/user-attachments/files/21567810/Design_.Cloud-Agnostic.Data.Warehouse.Support.in.Medplum.via.Apache.Iceberg.pdf)

Key features and use cases of this proposed design include:
* **Cloud-Agnosticism**: By using Apache Iceberg, which is an open table format for data lakes, Medplum can write data once and allow users to choose their preferred analytics engine, such as Amazon Redshift, Athena, Snowflake, Databricks, or Google BigQuery, to query the data. This prevents vendor lock-in and aligns with Medplum's philosophy of open standards.
* **Near Real-Time Replication**: The data pipeline will be streaming, with changes to FHIR data being queued and written to the Iceberg-based data lake within seconds. This ensures that analytical queries access fresh, up-to-date data.
* **HIPAA Compliance**: The data warehouse integration is designed to be HIPAA compliant. The Iceberg data lake will be in a secure, compliant environment, with all data encrypted at rest and in transit. Access controls and audit logging will be implemented, and Medplum's existing security features can be used for de-identification of data when needed.
* **Long-Term Data Retention**: High-volume, aging data like user sessions and audit events can be automatically offloaded from the primary Postgres store to the data lake for historical analysis and compliance, allowing the operational database to be trimmed for better performance.
* **Clinical and Business Intelligence Analytics**: The solution will enable complex SQL queries for population health and clinical analytics without impacting the performance of the operational FHIR store. Users will also be able to attach their preferred business intelligence (BI) tools and data warehouses to the Iceberg tables for live analytics and dashboard creation.

The architecture of the solution involves an event-driven, streaming pipeline. When a FHIR resource is created, updated, or deleted in the Medplum server, an event message is produced and sent to an ingestion queue. A dedicated Iceberg writer service then subscribes to this queue, batches the events, and writes the data to the corresponding Iceberg table in an object store, such as Amazon S3 or Google Cloud Storage. This process is designed to be low-latency and does not require a complex, external job scheduler. External query engines can then discover and query the Iceberg tables through a catalog service like AWS Glue Data Catalog.

**Comment by The-Alchemist** [comment]
Continuing this conversation which first started in October 2023.   I think it's worth summarizing a few things that have stayed the same, and things that have changed since this conversation started:


### New
* Apache Iceberg has more or less become a de facto data warehouse standard.  Every cloud (AWS, GCP, Azure) has a managed Iceberg offering, integrated into their suite of data science tools
   * for example, [AWS Redshift can query Iceberg tables](https://docs.aws.amazon.com/redshift/latest/dg/querying-iceberg.html)
   * [GCP's Iceberg support](https://cloud.google.com/blog/products/data-analytics/biglake-metastore-now-supports-iceberg-rest-catalog) with BigQuery
* Snowflake
   * rare company/product that companies are willing to leave the AWS ecosystem for
   * [proud of their Iceberg multi-cloud support](https://www.snowflake.com/en/product/features/iceberg/)

### Unchanged

* demand for data warehousing / "big data" analytics, especially health informatics, continues to grow
* we, as Medplum, want to support cloud-hosted _and_ self-hosted customers
* Medplum requires Node.js, Redis, PostgreSQL
  * we would _not_ like to increase the necessary deployment footprint

### Requirements / Restrictions

| Category | Self-hosted / cloud | Notes |
|----------|----------------------|--------|
| **Trust, access, and compliance** | **Hosted** (primary); **both** for boundaries | - For **cloud-hosted** customers, support **secure access to Medplum-operated analytics** with **strong access controls and auditing**, so **HIPAA / HITRUST** expectations are easier to meet (least privilege, traceability, reviewable access patterns).- Still document **where PHI lives** and **which system is authoritative** for each dataset. |
| **Tenant isolation** | **Hosted** | - Expect **row-level security (RLS)** (or an equivalent guardrail) so customer datasets **cannot cross tenant boundaries** in shared warehouse/query surfaces.<br>- Treat misconfiguration of RLS/policies as a **high-severity** class of bug. |
| **Operational shape (reads vs writes)** | **Hosted**; **both** as a principle | - **Decouple reads from writes**: analytics reads should not compete with transactional writes on the **API hot path**.<br>- **Some replication lag is acceptable** if it buys isolation, stability, and simpler scaling.<br>- Applies to “warehouse” whether it is Iceberg, Postgres replica, or something else. |
| **Architecture pragmatism (catalog vs “good enough”)** | **Hosted** | - A full **Iceberg catalog** may be **overkill** for many hosted use cases.<br>- A **read-only Postgres** connection (replica / analytics endpoint) can be a perfectly reasonable “warehouse” for reporting and internal tools **if** isolation, auditing, and performance limits are handled intentionally. |
| **Cost visibility (egress)** | **Hosted** | - Monitor **AWS egress** and large result transfers; **GB-scale** exports/queries should be **expected occasionally**, but watch for **runaway jobs** or accidental full-table pulls.<br>- Usually not a fundamental blocker—more of an **SRE/finops** guardrail. |
| **Portability and cloud lock-in** | **Both** | - Prefer patterns (formats, metadata, separation of storage and compute) that keep **migration and multi-cloud options real**, not theoretical.<br>- **Iceberg** is one useful “escape hatch,” but the requirement is **portability**, not a specific brand of catalog. |
| **Hybrid deployments** | **Self-hosted** (common); **both** possible | - **Ideal self-hosted flexibility:** run **local Postgres / Redis / etc.** for the app, while using a **cloud data warehouse** for analytics (pseudo-local control plane + pseudo-cloud analytics).<br>- Design exports/connectors so this split is **first-class**, not a hack. |
